src/main/java/uk/gov/hmcts/reform/bulkscanning/services/SasTokenGeneratorService.java

```diff
@@ -0,0 +1,95 @@
+import uk.gov.hmcts.reform.bulkscanning.exceptions.UnableToGenerateSasTokenException;
+
+import java.net.URISyntaxException;
+import java.security.InvalidKeyException;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.EnumSet;
+import java.util.GregorianCalendar;
+import java.util.List;
+import java.util.TimeZone;
+
+@EnableConfigurationProperties(AccessTokenConfiguration.class)
+@Service
+public class SasTokenGeneratorService {
+    private final CloudBlobClient cloudBlobClient;
+    private final StorageCredentials storageCredentials;
+    private final AccessTokenConfiguration accessTokenConfiguration;
+
+    private static final Logger log = LoggerFactory.getLogger(SasTokenGeneratorService.class);
+
+    public SasTokenGeneratorService(
+        CloudBlobClient cloudBlobClient,
+        StorageCredentials storageCredentials,
+        AccessTokenConfiguration accessTokenConfiguration
+    ) {
+        this.cloudBlobClient = cloudBlobClient;
+        this.storageCredentials = storageCredentials;
+        this.accessTokenConfiguration = accessTokenConfiguration;
+    }
+
+    public String generateSasToken(String serviceName) {
+        //Based on the service name container reference would be returned for specific service
+        try {
+            log.info("SAS Token request received for service {} ", serviceName);
+
+            CloudBlobContainer cloudBlobContainer = cloudBlobClient.getContainerReference(serviceName);
+
+            return cloudBlobContainer.generateSharedAccessSignature(createSharedAccessPolicy(serviceName), null);
+        } catch (URISyntaxException | StorageException | InvalidKeyException e) {
+            throw new UnableToGenerateSasTokenException(e);
+        }
+    }
+
+    private SharedAccessBlobPolicy createSharedAccessPolicy(String serviceName) {
+        AccessTokenConfiguration.TokenConfig config = getTokenConfigForService(serviceName);
+
+        Calendar cal = new GregorianCalendar(TimeZone.getTimeZone("UTC"));
+        cal.setTime(new Date());
+        cal.add(Calendar.SECOND, config.getValidity());
+
+        SharedAccessBlobPolicy policy = new SharedAccessBlobPolicy();
+        policy.setPermissions(getBlobPermissions(config));
+        policy.setSharedAccessExpiryTime(cal.getTime());
+
+        return policy;
+    }
+
+    private AccessTokenConfiguration.TokenConfig getTokenConfigForService(String serviceName) {
+        return accessTokenConfiguration.getServiceConfig().stream()
+            .filter(tokenConfig -> tokenConfig.getServiceName().equalsIgnoreCase(serviceName))
+            .findFirst()
+            .orElseThrow(
+                () -> new ServiceConfigNotFoundException("No service configuration found for service " + serviceName)
+            );
+    }
+
+    private EnumSet<SharedAccessBlobPermissions> getBlobPermissions(AccessTokenConfiguration.TokenConfig config) {
```

Comment:
Do we need to make permissions configurable? Can't we hard code this?

AFAIK we should only ever grant Write and List permissions; making this configurable introduces a risk of misconfiguration where we accidentally grant the token permissions to download the files.

Reply:
I initially thought to hardcode it but then thought when we want to download we will need to modify the code again so thought will make it configurable as it will then give us ability to configure it for download too.

Reply:
When we download we don't need to use a SAS, we can just use the normal Azure credentials & blob API.